<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#What-is-RPA?" data-toc-modified-id="What-is-RPA?-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>What is RPA?</a></span></li><li><span><a href="#What-types-of-tools-are-available?" data-toc-modified-id="What-types-of-tools-are-available?-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>What types of tools are available?</a></span></li><li><span><a href="#Areas-where-RPA-is-useful:" data-toc-modified-id="Areas-where-RPA-is-useful:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Areas where RPA is useful:</a></span></li><li><span><a href="#Example-use-case---TagUI-for-python" data-toc-modified-id="Example-use-case---TagUI-for-python-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Example use case - TagUI for python</a></span><ul class="toc-item"><li><span><a href="#Downloading-the-historical-price-data-for-bitcoin-and-GLDM-(gold-ETF)" data-toc-modified-id="Downloading-the-historical-price-data-for-bitcoin-and-GLDM-(gold-ETF)-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Downloading the historical price data for bitcoin and GLDM (gold ETF)</a></span><ul class="toc-item"><li><span><a href="#Download-bitcoin-and-GLDM-prices-from-yahoo-finance" data-toc-modified-id="Download-bitcoin-and-GLDM-prices-from-yahoo-finance-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Download bitcoin and GLDM prices from yahoo finance</a></span></li></ul></li></ul></li><li><span><a href="#Combine-RPA-with-other-python-packages:-example-use-case" data-toc-modified-id="Combine-RPA-with-other-python-packages:-example-use-case-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Combine RPA with other python packages: example use-case</a></span><ul class="toc-item"><li><span><a href="#Understanding-the-concepts" data-toc-modified-id="Understanding-the-concepts-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Understanding the concepts</a></span><ul class="toc-item"><li><span><a href="#Bitcoin-vs-GLDM" data-toc-modified-id="Bitcoin-vs-GLDM-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Bitcoin vs GLDM</a></span></li><li><span><a href="#open-vs-high-vs-low-vs-close-vs-adj-close" data-toc-modified-id="open-vs-high-vs-low-vs-close-vs-adj-close-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>open vs high vs low vs close vs adj close</a></span></li><li><span><a href="#Key-variables-to-use" data-toc-modified-id="Key-variables-to-use-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Key variables to use</a></span></li></ul></li><li><span><a href="#How-preprocessing-will-be-done:" data-toc-modified-id="How-preprocessing-will-be-done:-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>How preprocessing will be done:</a></span></li><li><span><a href="#Load-libraries" data-toc-modified-id="Load-libraries-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Load libraries</a></span></li><li><span><a href="#Load-dataset" data-toc-modified-id="Load-dataset-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Load dataset</a></span></li><li><span><a href="#Define-functions" data-toc-modified-id="Define-functions-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Define functions</a></span></li><li><span><a href="#bitcoin" data-toc-modified-id="bitcoin-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>bitcoin</a></span><ul class="toc-item"><li><span><a href="#Inspect-dataset" data-toc-modified-id="Inspect-dataset-5.6.1"><span class="toc-item-num">5.6.1&nbsp;&nbsp;</span>Inspect dataset</a></span></li><li><span><a href="#Missing-values" data-toc-modified-id="Missing-values-5.6.2"><span class="toc-item-num">5.6.2&nbsp;&nbsp;</span>Missing values</a></span></li><li><span><a href="#Unique-values" data-toc-modified-id="Unique-values-5.6.3"><span class="toc-item-num">5.6.3&nbsp;&nbsp;</span>Unique values</a></span></li><li><span><a href="#Duplicates" data-toc-modified-id="Duplicates-5.6.4"><span class="toc-item-num">5.6.4&nbsp;&nbsp;</span>Duplicates</a></span></li><li><span><a href="#Convert-data-types" data-toc-modified-id="Convert-data-types-5.6.5"><span class="toc-item-num">5.6.5&nbsp;&nbsp;</span>Convert data types</a></span></li><li><span><a href="#Set-date-as-row-index" data-toc-modified-id="Set-date-as-row-index-5.6.6"><span class="toc-item-num">5.6.6&nbsp;&nbsp;</span>Set date as row index</a></span></li><li><span><a href="#Check-date-range" data-toc-modified-id="Check-date-range-5.6.7"><span class="toc-item-num">5.6.7&nbsp;&nbsp;</span>Check date range</a></span></li><li><span><a href="#Rename-column-names" data-toc-modified-id="Rename-column-names-5.6.8"><span class="toc-item-num">5.6.8&nbsp;&nbsp;</span>Rename column names</a></span></li><li><span><a href="#Index-prices" data-toc-modified-id="Index-prices-5.6.9"><span class="toc-item-num">5.6.9&nbsp;&nbsp;</span>Index prices</a></span></li></ul></li><li><span><a href="#GLDM" data-toc-modified-id="GLDM-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>GLDM</a></span><ul class="toc-item"><li><span><a href="#Inspect-dataset" data-toc-modified-id="Inspect-dataset-5.7.1"><span class="toc-item-num">5.7.1&nbsp;&nbsp;</span>Inspect dataset</a></span></li><li><span><a href="#Missing-values" data-toc-modified-id="Missing-values-5.7.2"><span class="toc-item-num">5.7.2&nbsp;&nbsp;</span>Missing values</a></span></li><li><span><a href="#Unique-values" data-toc-modified-id="Unique-values-5.7.3"><span class="toc-item-num">5.7.3&nbsp;&nbsp;</span>Unique values</a></span></li><li><span><a href="#Duplicates" data-toc-modified-id="Duplicates-5.7.4"><span class="toc-item-num">5.7.4&nbsp;&nbsp;</span>Duplicates</a></span></li><li><span><a href="#Convert-data-types" data-toc-modified-id="Convert-data-types-5.7.5"><span class="toc-item-num">5.7.5&nbsp;&nbsp;</span>Convert data types</a></span></li><li><span><a href="#Set-date-as-row-index" data-toc-modified-id="Set-date-as-row-index-5.7.6"><span class="toc-item-num">5.7.6&nbsp;&nbsp;</span>Set date as row index</a></span></li><li><span><a href="#Check-date-range" data-toc-modified-id="Check-date-range-5.7.7"><span class="toc-item-num">5.7.7&nbsp;&nbsp;</span>Check date range</a></span></li><li><span><a href="#Rename-column-names" data-toc-modified-id="Rename-column-names-5.7.8"><span class="toc-item-num">5.7.8&nbsp;&nbsp;</span>Rename column names</a></span></li><li><span><a href="#Index-prices" data-toc-modified-id="Index-prices-5.7.9"><span class="toc-item-num">5.7.9&nbsp;&nbsp;</span>Index prices</a></span></li></ul></li><li><span><a href="#Merge-indexed-bitcoin-with-GLDM" data-toc-modified-id="Merge-indexed-bitcoin-with-GLDM-5.8"><span class="toc-item-num">5.8&nbsp;&nbsp;</span>Merge indexed bitcoin with GLDM</a></span><ul class="toc-item"><li><span><a href="#Dealing-with-the-NA-values-due-to-missing-weekend-data-for-GLDM" data-toc-modified-id="Dealing-with-the-NA-values-due-to-missing-weekend-data-for-GLDM-5.8.1"><span class="toc-item-num">5.8.1&nbsp;&nbsp;</span>Dealing with the NA values due to missing weekend data for GLDM</a></span></li></ul></li><li><span><a href="#Get-dataframe-for-linechart" data-toc-modified-id="Get-dataframe-for-linechart-5.9"><span class="toc-item-num">5.9&nbsp;&nbsp;</span>Get dataframe for linechart</a></span></li><li><span><a href="#Load-dataframe-as-excel-file" data-toc-modified-id="Load-dataframe-as-excel-file-5.10"><span class="toc-item-num">5.10&nbsp;&nbsp;</span>Load dataframe as excel file</a></span><ul class="toc-item"><li><span><a href="#Define-ExcelChart-class" data-toc-modified-id="Define-ExcelChart-class-5.10.1"><span class="toc-item-num">5.10.1&nbsp;&nbsp;</span>Define ExcelChart class</a></span></li></ul></li></ul></li><li><span><a href="#Create-new-excel-sheet-with-line-chart,-append-raw-dataframes-as-separate-worksheets,-then-save-the-excel-workbook" data-toc-modified-id="Create-new-excel-sheet-with-line-chart,-append-raw-dataframes-as-separate-worksheets,-then-save-the-excel-workbook-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Create new excel sheet with line chart, append raw dataframes as separate worksheets, then save the excel workbook</a></span></li><li><span><a href="#Further-ideas" data-toc-modified-id="Further-ideas-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Further ideas</a></span></li></ul></div>

# What is RPA?

Robotic Process Automation, or [RPA](https://en.wikipedia.org/wiki/Robotic_process_automation) is a type of software technology that helps to automate repetitive tasks. 

# What types of tools are available?

There are many tools available in the market. These include paid software such as [UiPath](https://www.uipath.com/) and free tools such as [TagUI](https://aisingapore.org/tagui/)

# Areas where RPA is useful: 

Any repetitive computer-related task that is based on known rules, and which can be broken down or operationalised into a series of steps  

# Example use case - TagUI for python

To demonstrate how easy it is to use TagUI, we can use it to download historial price data for bitcoin and GLDM (gold ETF) from the web. Although this is less efficient than using other known python packages for web scraping, the same concepts can also be applied when automating well-defined repetitive tasks within legacy software systems. In those instances where automation is done from the user's perspective, computer vision can be used to assist in the automation process. 

For the example below, we will be using xpaths to idenfity elements within the yahoo finance website. 

TagUI can be installed from [here](https://tagui.readthedocs.io/en/latest/setup.html)  
- The API documentation for TagUI for python can be found [here](https://github.com/tebelorg/RPA-Python)  

## Downloading the historical price data for bitcoin and GLDM (gold ETF)

Task: Download bitcoin and GLDM (gold ETF) price data from yahoo finance using TagUI in python
  - Date range: June 1, 2019 to today  

Break down the task into a series of steps:

### Download bitcoin and GLDM prices from yahoo finance

In [1]:
from datetime import datetime 
from pathlib import Path
import os # used to change directory
import rpa as r

TODAY = str(datetime.today().strftime('%d/%m/%Y'))

search_terms = ['bitcoin', 'gldm']
start_date = '01/06/2019'
end_date = TODAY
raw_csv_folder = Path.cwd().parents[0]/'data/raw'

def download_csv(
    raw_csv_folder: str,
    search_terms: list,
    start_date: str,
    end_date: str
):
    """Download csv files of historical price data for specified financial assets from yahoo finance
    - xpaths are based on yahoo finance website

    Args:
        raw_csv_folder (str): absolute file path to folder to save raw csv files
        search_terms (list): list of financial assets to search for and download csv of historical prices
        start_date (str): start date of historical price data
        end_date (str): end date of historical price data.
    """
    # change working directory as the downloaded csv files are saved into current working dir
    os.chdir(raw_csv_folder)

    # instantiate rpa 
    r.init()

    # go to yahoo finance to obtain price data
    r.url('https://finance.yahoo.com')

    # wait for the page to load
    r.wait(3)

    for search_term in search_terms:
        
        # use yahoo finance search bar
        r.type('//*[@id="yfin-usr-qry"]', search_term + '[enter]')

        r.wait(1)
        
        # go to historical data link
        r.click('//*[@data-test="HISTORICAL_DATA"]')
        
        # click on date range
        r.click('//*[@id="Col1-1-HistoricalDataTable-Proxy"]/section/div[1]/div[1]/div[1]/div/div/div/span')
        
        # wait for dropdown menu to pop up before entering date
        r.wait(2)
        
        # input starting date range
        r.type('//*[@id="dropdown-menu"]/div/div[1]/input', start_date)
        
        # input end date
        r.type('//*[@id="dropdown-menu"]/div/div[2]/input', end_date)
        
        # click 'done' button in dropdown menu
        r.click('//*[@id="dropdown-menu"]/div/div[3]/button[1]/span')
        
        # click apply button to apply date range
        r.click('//*[@id="Col1-1-HistoricalDataTable-Proxy"]/section/div[1]/div[1]/button/span')
        
        # download the csv file 
        r.click('//*[@id="Col1-1-HistoricalDataTable-Proxy"]/section/div[1]/div[2]/span[2]/a/span')
        
    r.close()

In [2]:
download_csv(
    raw_csv_folder = raw_csv_folder,
    search_terms = search_terms,
    start_date = start_date,
    end_date = end_date
)

# Combine RPA with other python packages: example use-case

We can also combine RPA with other python libraries like pandas and openpyxl. 
For example, with the downloaded bitcoin and GLDM (gold ETF) csv, we can then merge both datasets and create a combined line chart in excel that displays the indexed prices (indexed to 1/6/2019 price) for bitcoin and GLDM in the same excel chart. 

Here's a breakdown of the steps: 
1. Process both sets of price data: 
  - index bitcoin and GLDM prices based on 1/6/2019 price, with all other prices relative to that price ('indexing to 100')
2. Merge both sets of price data into the same spreadsheet 
3. Create a line chart showing the daily indexed price history from June 1, 2019 to today  
4. Save output as excel file with line chart  

However, before we can do this, we should understand the underlying concepts related to bitcoin and GLDM, and first explore the respective datasets. 

## Understanding the concepts

To understand how we should be processing both csv files, we should try to understand the terms and concepts related to bitcoin and GLDM

### Bitcoin vs GLDM

- **Bitcoin**: cryptocurrency
  - generated through the process of mining 
  - currently trading on 9598 active markets
  - trades 24/7  
  - can be bought through cryptocurrency exchanges
  - source of yahoo finance historical price data is actually from [CoinMarketCap](https://coinmarketcap.com)  
  - from [CoinMarketCap historical data](https://coinmarketcap.com/currencies/bitcoin/historical-data) the price at open/close is actually based on Coordinated Universal Time (UTC)    
  - from yahoo finance, the ['Close'](https://finance.yahoo.com/quote/BTC-USD/history?p=BTC-USD) price has been adjusted for splits, and the [Adjusted close price](https://finance.yahoo.com/quote/BTC-USD/history?p=BTC-USD) has been adjusted for both dividends and splits.  
  
---

- **GLDM** Gold ETF (mini): Exchange Traded Fund to track price of gold  
  - trades on NYSEArca exchange  
  - only trades Mondays - Fridays excluding US public holidays  
  - open/close prices based on trading price at market open/just before market closes at Eastern Daylight Time (EDT)  
  - similar to bitcoin, the ['Close'](https://finance.yahoo.com/quote/GLDM/history?p=GLDM) price has been adjusted for splits, while the ['Adjusted Close Price'](https://finance.yahoo.com/quote/GLDM/history?p=GLDM) has been adjusted for both dividends and splits.  

### open vs high vs low vs close vs adj close 

- open 
  - price at which a stock started trading when the opening bell rang
  
- close 
  - price of a stock when the stock exchanged has closed for the day (last buy-sell order executed between 2 traders) 
    - considered the reference point for any time frame 
    
- high
  - the highest price at which a stock traded during a period (since this is a daily chart, it would be during the day)  
  
- low 
  - lowest price at which a stock traded during the same time period (i.e. intraday low)  
  
- volume 
  - total number of shares traded in the stock/security over a timer period (i.e. during the day for this set of data)  

- ['Adjusted Close Price'](https://finance.yahoo.com/quote/GLDM/history?p=GLDM)  
  - closing price adjusted for both dividends and splits  
    - stock split: corporate action in which a company increases the number of its outstanding shares by issuing more shares to current shareholders  
    - dividend: distribution of a company's earnings to a class of its shareholders  
    - why it matters: stock splits and dividend payouts affect the stock's price and value of the stock  
    - this would be a more useful variable to use when compring the performance between asset classes over the long term  

sources: 
[Analyzing Alpha](https://analyzingalpha.com/open-high-low-close-stocks), [Investopedia](https://www.investopedia.com/terms/a/adjusted_closing_price.asp)  

### Key variables to use

By [indexing](https://www.dallasfed.org/research/basics/indexing.aspx) the data to a common start date (i.e. in this case 1 June 2019), we are trying to normalise the data to a common starting point and observing how the price of bitcoin and GLDM change over time relative to each other.  
- However, 1 June 2019 is a Saturday and GLDM does not trade on a Saturday  
- There are 2 alternatives to facilitate comparison between the historical returns of bitcoin vs GLDM   
  1. index to 3 June 2019 (Monday) instead of 1 June 2019 as GLDM would be NaN on 1 June 2019  
    - but the problem is that we will still have NA values for the other weekend dates  
  2. index price data separately first, then combine. Then fill the missing values with a combination of forward fill and backfill. (see below)


**Dates**
  - Dates for GLDM will only be available from Monday - Fridays excluding US Public holidays  
  - While dates for Bitcoin will be available for every single day 
- When we merge both sets of price data, we can choose to merge on bitcoin or GLDM  
- if we choose to merge on GLDM, then we will be losing important information on Bitcoin price changes over the weekend when GLDM is not trading 
- By choosing to merge on bitcoin, there will be missing values for the GLDM data  
    - for the missing GLDM values, we can propagate last valid observation forward to next valid (forward fill) as that would reflect the price from Friday 'close' to Monday (i.e. when trading resumes)  

## How preprocessing will be done: 
Based on the concepts discussed earlier, preprocessing will be done as per the following: 

- inspect each dataset separately 
- convert data types 
- rename column names according to the asset type (i.e. bitcoin or gldm)
- index prices to start date (1 June 2019) 
- merge indexed GLDM to bitcoin on date column, use adjusted closing price for the chart
- NA values:
  - for GLDM chart values (adjusted closing price): forward fill the weekend dates (i.e. fill Sat and Sun data with Friday's closing price)  
  - then backfill NA values for first 2 dates (since we do not have the data) as the start date (1 June 2019) falls on a weekend.
- we will still keep all raw variables (eg. open/close/high/low) and indexed variables in separate spreadsheets in case the user needs to create candlestick charts or needs to access the raw dataset     


## Load libraries

In [3]:
from datetime import datetime
import glob
from openpyxl import Workbook
from openpyxl.utils.dataframe import dataframe_to_rows
from openpyxl.chart import LineChart, Reference
from openpyxl.chart.axis import DateAxis
import pandas as pd
from pathlib import Path

## Load dataset

In [4]:
all_filenames = [name for name in glob.glob(f'*.{"csv"}')]
all_filenames

['BTC-USD.csv', 'GLDM.csv']

In [5]:
bitcoin = pd.read_csv('BTC-USD.csv')
gldm = pd.read_csv('GLDM.csv')

## Define functions

In [6]:
def check_missing(df):
    print(f"Number of missing or NA values for each column out of {len(df)} rows:\
          \n {df.isna().sum()}")
    print('#'*40)

    print(f"Percentage of missing or NA values for each column out of {len(df)} rows:\
          \n {round(df.isna().sum()/len(df)*100, 2)}")
    
def check_unique(df):
    print(f'Number of unique values for each column out of a total of {len(df)} rows: \n {df.nunique()}')
    
def rename_col(asset_name: str, df: pd.DataFrame):
    """Rename column names based on financial asset name

    Args:
        asset_name (str): financial asset
        df (pd.DataFrame): pandas dataframe containing price data 

    Returns:
        df (pd.DataFrame): pandas dataframe with columns renamed
    """

    newcol=[]

    for col in df.columns.tolist():
        newcol.append(col + '_' + asset_name)

    df.columns = newcol

    return df

def index_prices(df: pd.DataFrame):
    """Index price data according to the value at the start date

    Args:
        df (pd.DataFrame): pandas dataframe containing price data

    Returns:
        df_indexed (pd.DataFrame): pandas dataframe with price data indexed
    """

    # Set base prices row and columns
    base_prices = df.iloc[0, :]

    # Indexed prices
    df_indexed = df.div(base_prices/100, axis =1)

    return df_indexed

## bitcoin

### Inspect dataset

In [7]:
print(f"Number of rows and columns: {bitcoin.shape}")

Number of rows and columns: (865, 7)


In [8]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 865 entries, 0 to 864
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       865 non-null    object 
 1   Open       861 non-null    float64
 2   High       861 non-null    float64
 3   Low        861 non-null    float64
 4   Close      861 non-null    float64
 5   Adj Close  861 non-null    float64
 6   Volume     861 non-null    float64
dtypes: float64(6), object(1)
memory usage: 47.4+ KB


In [9]:
bitcoin.head(5)

Date         Open         High          Low        Close  \
0  2019-06-01  8573.839844  8625.600586  8481.578125  8564.016602   
1  2019-06-02  8565.473633  8809.303711  8561.235352  8742.958008   
2  2019-06-03  8741.747070  8743.500000  8204.185547  8208.995117   
3  2019-06-04  8210.985352  8210.985352  7564.488770  7707.770996   
4  2019-06-05  7704.343262  7901.849121  7668.668457  7824.231445   

     Adj Close        Volume  
0  8564.016602  2.248830e+10  
1  8742.958008  2.026622e+10  
2  8208.995117  2.200451e+10  
3  7707.770996  2.460973e+10  
4  7824.231445  2.176092e+10

In [10]:
bitcoin.tail(5)

Date          Open          High           Low         Close  \
860  2021-10-08  53802.144531  55922.980469  53688.054688  53967.847656   
861  2021-10-09  53929.781250  55397.945313  53735.144531  54968.222656   
862  2021-10-10  54952.820313  56401.304688  54264.257813  54771.578125   
863  2021-10-11  54734.125000  57793.039063  54519.765625  57484.789063   
864  2021-10-12  57413.859375  57413.859375  56526.679688  57010.562500   

        Adj Close        Volume  
860  53967.847656  3.480087e+10  
861  54968.222656  3.249121e+10  
862  54771.578125  3.952779e+10  
863  57484.789063  4.263733e+10  
864  57010.562500  4.110214e+10

### Missing values 

In [11]:
check_missing(bitcoin)

Number of missing or NA values for each column out of 865 rows:          
 Date         0
Open         4
High         4
Low          4
Close        4
Adj Close    4
Volume       4
dtype: int64
########################################
Percentage of missing or NA values for each column out of 865 rows:          
 Date         0.00
Open         0.46
High         0.46
Low          0.46
Close        0.46
Adj Close    0.46
Volume       0.46
dtype: float64


only 4 rows out of 861 rows are missing - minimal impact on analyses. Check if they are important dates:

In [12]:
bitcoin.loc[bitcoin['Adj Close'].isnull(),:]

Date  Open  High  Low  Close  Adj Close  Volume
321  2020-04-17   NaN   NaN  NaN    NaN        NaN     NaN
496  2020-10-09   NaN   NaN  NaN    NaN        NaN     NaN
499  2020-10-12   NaN   NaN  NaN    NaN        NaN     NaN
500  2020-10-13   NaN   NaN  NaN    NaN        NaN     NaN

Looks like perhaps the website was down for maintenance those few days. 
Since there's only 4 rows missing (0.46%), it will have minimal impact on subsequent analyses. 

Since bitcoin is an asset with very volatile price fluctuations, we **should not** fill those NA values as those filled values might not represent the value of central tendency for that asset/variable at that point in time. 
The similar concept will also be applied to other cryptocurrencies - since they are volatile price assets. 

### Unique values

In [13]:
check_unique(bitcoin)

Number of unique values for each column out of a total of 865 rows: 
 Date         865
Open         861
High         861
Low          861
Close        861
Adj Close    861
Volume       861
dtype: int64


### Duplicates

In [14]:
print(f"There are {bitcoin.duplicated().sum()} complete duplicate rows in this dataset")

There are 0 complete duplicate rows in this dataset


### Convert data types 

In [15]:
bitcoin.Date = pd.to_datetime(
    bitcoin.Date, 
    format= '%Y-%m-%d'
)

In [16]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 865 entries, 0 to 864
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       865 non-null    datetime64[ns]
 1   Open       861 non-null    float64       
 2   High       861 non-null    float64       
 3   Low        861 non-null    float64       
 4   Close      861 non-null    float64       
 5   Adj Close  861 non-null    float64       
 6   Volume     861 non-null    float64       
dtypes: datetime64[ns](1), float64(6)
memory usage: 47.4 KB


### Set date as row index

In [17]:
bitcoin = bitcoin.set_index('Date')

### Check date range

In [18]:
str(bitcoin.index.date.min())

'2019-06-01'

In [19]:
str(bitcoin.index.date.max())

'2021-10-12'

### Rename column names

In [20]:
bitcoin = rename_col('BTC', bitcoin)

In [21]:
# Checking that it was done correctly
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 865 entries, 2019-06-01 to 2021-10-12
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Open_BTC       861 non-null    float64
 1   High_BTC       861 non-null    float64
 2   Low_BTC        861 non-null    float64
 3   Close_BTC      861 non-null    float64
 4   Adj Close_BTC  861 non-null    float64
 5   Volume_BTC     861 non-null    float64
dtypes: float64(6)
memory usage: 47.3 KB


### Index prices

In [22]:
bitcoin_indexed = index_prices(bitcoin)

In [23]:
# Checking the indexed dataframe
bitcoin_indexed

Open_BTC    High_BTC     Low_BTC   Close_BTC  Adj Close_BTC  \
Date                                                                        
2019-06-01  100.000000  100.000000  100.000000  100.000000     100.000000   
2019-06-02   99.902422  102.129743  100.939179  102.089457     102.089457   
2019-06-03  101.958367  101.366855   96.729470   95.854498      95.854498   
2019-06-04   95.767888   95.193202   89.187279   90.001822      90.001822   
2019-06-05   89.858726   91.609263   90.415585   91.361703      91.361703   
...                ...         ...         ...         ...            ...   
2021-10-08  627.515157  648.337236  632.996052  630.169816     630.169816   
2021-10-09  629.003833  642.250296  633.551254  641.850959     641.850959   
2021-10-10  640.935932  653.882638  639.789636  639.554787     639.554787   
2021-10-11  638.385204  670.017566  642.802139  671.236310     671.236310   
2021-10-12  669.639980  665.621585  666.464175  665.698879     665.698879   

            Volume_BTC  
Date                    
2019-06-01  100.000000  
2019-06-02   90.118919  
2019-06-03   97.848695  
2019-06-04  109.433473  
2019-06-05   96.765518  
...                ...  
2021-10-08  154.750997  
2021-10-09  144.480491  
2021-10-10  175.770450  
2021-10-11  189.597813  
2021-10-12  182.771206  

[865 rows x 6 columns]

Double checking that it was done correctly:

In [24]:
# Comparing with the original non-indexed dataframe
bitcoin

Open_BTC      High_BTC       Low_BTC     Close_BTC  \
Date                                                                 
2019-06-01   8573.839844   8625.600586   8481.578125   8564.016602   
2019-06-02   8565.473633   8809.303711   8561.235352   8742.958008   
2019-06-03   8741.747070   8743.500000   8204.185547   8208.995117   
2019-06-04   8210.985352   8210.985352   7564.488770   7707.770996   
2019-06-05   7704.343262   7901.849121   7668.668457   7824.231445   
...                  ...           ...           ...           ...   
2021-10-08  53802.144531  55922.980469  53688.054688  53967.847656   
2021-10-09  53929.781250  55397.945313  53735.144531  54968.222656   
2021-10-10  54952.820313  56401.304688  54264.257813  54771.578125   
2021-10-11  54734.125000  57793.039063  54519.765625  57484.789063   
2021-10-12  57413.859375  57413.859375  56526.679688  57010.562500   

            Adj Close_BTC    Volume_BTC  
Date                                     
2019-06-01    8564.016602  2.248830e+10  
2019-06-02    8742.958008  2.026622e+10  
2019-06-03    8208.995117  2.200451e+10  
2019-06-04    7707.770996  2.460973e+10  
2019-06-05    7824.231445  2.176092e+10  
...                   ...           ...  
2021-10-08   53967.847656  3.480087e+10  
2021-10-09   54968.222656  3.249121e+10  
2021-10-10   54771.578125  3.952779e+10  
2021-10-11   57484.789063  4.263733e+10  
2021-10-12   57010.562500  4.110214e+10  

[865 rows x 6 columns]

Check base price (price on 2019-06-01) for 'Adj Close_BTC':

In [25]:
base_price = bitcoin.loc['2019-06-01']['Adj Close_BTC']
base_price

8564.016602

Check indexed price on 2019-06-02:

In [26]:
bitcoin.loc['2019-06-02']['Adj Close_BTC'] / base_price*100

102.08945655194329

In [27]:
bitcoin_indexed.loc['2019-06-02']['Adj Close_BTC']

102.08945655194329

Check indexed price on 2021-10-07: 

In [28]:
bitcoin.loc['2021-10-07']['Adj Close_BTC'] / base_price*100

628.2797765996204

In [29]:
bitcoin_indexed.loc['2021-10-07']['Adj Close_BTC']

628.2797765996204

In [30]:
bitcoin_indexed.loc['2021-10-07']['Adj Close_BTC'] == bitcoin.loc['2021-10-07']['Adj Close_BTC'] / base_price*100

True

## GLDM

### Inspect dataset

In [31]:
print(f"Number of rows and columns: {gldm.shape}")

Number of rows and columns: (596, 7)


In [32]:
gldm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596 entries, 0 to 595
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       596 non-null    object 
 1   Open       596 non-null    float64
 2   High       596 non-null    float64
 3   Low        596 non-null    float64
 4   Close      596 non-null    float64
 5   Adj Close  596 non-null    float64
 6   Volume     596 non-null    int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 32.7+ KB


In [33]:
gldm.head(5)

Date   Open   High    Low  Close  Adj Close   Volume
0  2019-06-03  13.13  13.26  13.11  13.23      13.23  1701600
1  2019-06-04  13.21  13.26  13.18  13.24      13.24  1044700
2  2019-06-05  13.35  13.37  13.26  13.27      13.27  1233200
3  2019-06-06  13.33  13.38  13.31  13.31      13.31   535500
4  2019-06-07  13.41  13.46  13.37  13.39      13.39  1170800

In [34]:
gldm.tail(5)

Date       Open       High        Low      Close  Adj Close  \
591  2021-10-05  17.440001  17.528000  17.389999  17.500000  17.500000   
592  2021-10-06  17.450001  17.548000  17.450001  17.530001  17.530001   
593  2021-10-07  17.450001  17.525000  17.430000  17.459999  17.459999   
594  2021-10-08  17.690001  17.690001  17.455999  17.469999  17.469999   
595  2021-10-11  17.450001  17.497999  17.430000  17.430000  17.430000   

      Volume  
591  2195700  
592  2241800  
593  1277600  
594  2232900  
595   959100

### Missing values 

In [35]:
check_missing(gldm)

Number of missing or NA values for each column out of 596 rows:          
 Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64
########################################
Percentage of missing or NA values for each column out of 596 rows:          
 Date         0.0
Open         0.0
High         0.0
Low          0.0
Close        0.0
Adj Close    0.0
Volume       0.0
dtype: float64


### Unique values

In [36]:
check_unique(gldm)

Number of unique values for each column out of a total of 596 rows: 
 Date         596
Open         352
High         446
Low          432
Close        362
Adj Close    362
Volume       591
dtype: int64


### Duplicates

In [37]:
print(f"There are {gldm.duplicated().sum()} complete duplicate rows in this dataset")

There are 0 complete duplicate rows in this dataset


### Convert data types 

In [38]:
gldm.Date = pd.to_datetime(
    gldm.Date, 
    format= '%Y-%m-%d'
)

In [39]:
# Checking that it was done correctly 
gldm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596 entries, 0 to 595
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       596 non-null    datetime64[ns]
 1   Open       596 non-null    float64       
 2   High       596 non-null    float64       
 3   Low        596 non-null    float64       
 4   Close      596 non-null    float64       
 5   Adj Close  596 non-null    float64       
 6   Volume     596 non-null    int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 32.7 KB


### Set date as row index

In [40]:
gldm = gldm.set_index('Date')

### Check date range

In [41]:
str(gldm.index.date.min())

'2019-06-03'

In [42]:
str(gldm.index.date.max())

'2021-10-11'

### Rename column names

In [43]:
gldm = rename_col('GLDM', gldm)

In [44]:
gldm.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 596 entries, 2019-06-03 to 2021-10-11
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Open_GLDM       596 non-null    float64
 1   High_GLDM       596 non-null    float64
 2   Low_GLDM        596 non-null    float64
 3   Close_GLDM      596 non-null    float64
 4   Adj Close_GLDM  596 non-null    float64
 5   Volume_GLDM     596 non-null    int64  
dtypes: float64(5), int64(1)
memory usage: 32.6 KB


### Index prices

In [45]:
gldm_indexed = index_prices(gldm)

In [46]:
# Checking the indexed dataframe
gldm_indexed

Open_GLDM   High_GLDM    Low_GLDM  Close_GLDM  Adj Close_GLDM  \
Date                                                                         
2019-06-03  100.000000  100.000000  100.000000  100.000000      100.000000   
2019-06-04  100.609292  100.000000  100.533944  100.075586      100.075586   
2019-06-05  101.675552  100.829563  101.144165  100.302343      100.302343   
2019-06-06  101.523229  100.904977  101.525553  100.604686      100.604686   
2019-06-07  102.132521  101.508296  101.983219  101.209373      101.209373   
...                ...         ...         ...         ...             ...   
2021-10-05  132.825598  132.187029  132.646827  132.275132      132.275132   
2021-10-06  132.901759  132.337858  133.104508  132.501897      132.501897   
2021-10-07  132.901759  132.164404  132.951945  131.972782      131.972782   
2021-10-08  134.729634  133.408756  133.150259  132.048367      132.048367   
2021-10-11  132.901759  131.960777  132.951945  131.746032      131.746032   

            Volume_GLDM  
Date                     
2019-06-03   100.000000  
2019-06-04    61.395157  
2019-06-05    72.472967  
2019-06-06    31.470381  
2019-06-07    68.805830  
...                 ...  
2021-10-05   129.037377  
2021-10-06   131.746591  
2021-10-07    75.082276  
2021-10-08   131.223554  
2021-10-11    56.364598  

[596 rows x 6 columns]

In [47]:
# Comparing with the original non-indexed dataframe
gldm

Open_GLDM  High_GLDM   Low_GLDM  Close_GLDM  Adj Close_GLDM  \
Date                                                                      
2019-06-03  13.130000  13.260000  13.110000   13.230000       13.230000   
2019-06-04  13.210000  13.260000  13.180000   13.240000       13.240000   
2019-06-05  13.350000  13.370000  13.260000   13.270000       13.270000   
2019-06-06  13.330000  13.380000  13.310000   13.310000       13.310000   
2019-06-07  13.410000  13.460000  13.370000   13.390000       13.390000   
...               ...        ...        ...         ...             ...   
2021-10-05  17.440001  17.528000  17.389999   17.500000       17.500000   
2021-10-06  17.450001  17.548000  17.450001   17.530001       17.530001   
2021-10-07  17.450001  17.525000  17.430000   17.459999       17.459999   
2021-10-08  17.690001  17.690001  17.455999   17.469999       17.469999   
2021-10-11  17.450001  17.497999  17.430000   17.430000       17.430000   

            Volume_GLDM  
Date                     
2019-06-03      1701600  
2019-06-04      1044700  
2019-06-05      1233200  
2019-06-06       535500  
2019-06-07      1170800  
...                 ...  
2021-10-05      2195700  
2021-10-06      2241800  
2021-10-07      1277600  
2021-10-08      2232900  
2021-10-11       959100  

[596 rows x 6 columns]

Check base price (price on 2019-06-03) for 'Adj Close_GLDM':

In [48]:
base_price = gldm.loc['2019-06-03']['Adj Close_GLDM']
base_price

13.23

Check indexed price on 2019-06-04:

In [49]:
gldm.loc['2019-06-04']['Adj Close_GLDM'] / base_price*100

100.0755857898715

In [50]:
gldm_indexed.loc['2019-06-04']['Adj Close_GLDM']

100.07558578987151

Check indexed price on 2021-10-07: 

In [51]:
gldm.loc['2021-10-07']['Adj Close_GLDM'] / base_price*100

131.97278155706726

In [52]:
gldm_indexed.loc['2021-10-07']['Adj Close_GLDM']

131.97278155706726

In [53]:
gldm.loc['2021-10-07']['Adj Close_GLDM'] / base_price*100 == gldm_indexed.loc['2021-10-07']['Adj Close_GLDM']

True

## Merge indexed bitcoin with GLDM

In [54]:
df = pd.merge(
    bitcoin_indexed, 
    gldm_indexed, 
    how='left', 
    left_index=True, 
    right_index=True
)

In [55]:
# Double checking dataframe after the merge
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 865 entries, 2019-06-01 to 2021-10-12
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Open_BTC        861 non-null    float64
 1   High_BTC        861 non-null    float64
 2   Low_BTC         861 non-null    float64
 3   Close_BTC       861 non-null    float64
 4   Adj Close_BTC   861 non-null    float64
 5   Volume_BTC      861 non-null    float64
 6   Open_GLDM       596 non-null    float64
 7   High_GLDM       596 non-null    float64
 8   Low_GLDM        596 non-null    float64
 9   Close_GLDM      596 non-null    float64
 10  Adj Close_GLDM  596 non-null    float64
 11  Volume_GLDM     596 non-null    float64
dtypes: float64(12)
memory usage: 120.1 KB


In [56]:
# Double checking the dataframe 
df

Open_BTC    High_BTC     Low_BTC   Close_BTC  Adj Close_BTC  \
Date                                                                        
2019-06-01  100.000000  100.000000  100.000000  100.000000     100.000000   
2019-06-02   99.902422  102.129743  100.939179  102.089457     102.089457   
2019-06-03  101.958367  101.366855   96.729470   95.854498      95.854498   
2019-06-04   95.767888   95.193202   89.187279   90.001822      90.001822   
2019-06-05   89.858726   91.609263   90.415585   91.361703      91.361703   
...                ...         ...         ...         ...            ...   
2021-10-08  627.515157  648.337236  632.996052  630.169816     630.169816   
2021-10-09  629.003833  642.250296  633.551254  641.850959     641.850959   
2021-10-10  640.935932  653.882638  639.789636  639.554787     639.554787   
2021-10-11  638.385204  670.017566  642.802139  671.236310     671.236310   
2021-10-12  669.639980  665.621585  666.464175  665.698879     665.698879   

            Volume_BTC   Open_GLDM   High_GLDM    Low_GLDM  Close_GLDM  \
Date                                                                     
2019-06-01  100.000000         NaN         NaN         NaN         NaN   
2019-06-02   90.118919         NaN         NaN         NaN         NaN   
2019-06-03   97.848695  100.000000  100.000000  100.000000  100.000000   
2019-06-04  109.433473  100.609292  100.000000  100.533944  100.075586   
2019-06-05   96.765518  101.675552  100.829563  101.144165  100.302343   
...                ...         ...         ...         ...         ...   
2021-10-08  154.750997  134.729634  133.408756  133.150259  132.048367   
2021-10-09  144.480491         NaN         NaN         NaN         NaN   
2021-10-10  175.770450         NaN         NaN         NaN         NaN   
2021-10-11  189.597813  132.901759  131.960777  132.951945  131.746032   
2021-10-12  182.771206         NaN         NaN         NaN         NaN   

            Adj Close_GLDM  Volume_GLDM  
Date                                     
2019-06-01             NaN          NaN  
2019-06-02             NaN          NaN  
2019-06-03      100.000000   100.000000  
2019-06-04      100.075586    61.395157  
2019-06-05      100.302343    72.472967  
...                    ...          ...  
2021-10-08      132.048367   131.223554  
2021-10-09             NaN          NaN  
2021-10-10             NaN          NaN  
2021-10-11      131.746032    56.364598  
2021-10-12             NaN          NaN  

[865 rows x 12 columns]

### Dealing with the NA values due to missing weekend data for GLDM

**1. Mising GLDM values:**  

As discussed earlier, by choosing to merge on bitcoin, there will be missing values for the GLDM data for the weekend dates as GLDM does not trade on weekends. 
- for GLDM chart values (adjusted closing price): we will backfill NA values for first 2 dates (since we do not have the data as 1 June 2019 is a Saturday), 
- for the other GLDM missing values, we can propagate last valid observation forward to next valid (forward fill) as that would reflect the price from Friday 'close' to Monday (i.e. when trading resumes)  

**2. Missing bitcoin values:**
- Since bitcoin is an asset with very volatile price fluctuations, we will not fill those NA values as those filled values might not represent the value of central tendency for that asset/variable at that point in time. We will just leave the 4 NA values for bitcoin as it is. 

In [57]:
# Double checking missing values: 
df.isnull().sum()

Open_BTC            4
High_BTC            4
Low_BTC             4
Close_BTC           4
Adj Close_BTC       4
Volume_BTC          4
Open_GLDM         269
High_GLDM         269
Low_GLDM          269
Close_GLDM        269
Adj Close_GLDM    269
Volume_GLDM       269
dtype: int64

In [58]:
gldm_cols = [col for col in df.columns.tolist() if 'GLDM' in col]
gldm_cols

['Open_GLDM',
 'High_GLDM',
 'Low_GLDM',
 'Close_GLDM',
 'Adj Close_GLDM',
 'Volume_GLDM']

In [59]:
df[gldm_cols] = df[gldm_cols].fillna(method='ffill')

In [60]:
# Checking NAs after forward fill
df.isnull().sum()

Open_BTC          4
High_BTC          4
Low_BTC           4
Close_BTC         4
Adj Close_BTC     4
Volume_BTC        4
Open_GLDM         2
High_GLDM         2
Low_GLDM          2
Close_GLDM        2
Adj Close_GLDM    2
Volume_GLDM       2
dtype: int64

In [61]:
# Double checking dataframe
df

Open_BTC    High_BTC     Low_BTC   Close_BTC  Adj Close_BTC  \
Date                                                                        
2019-06-01  100.000000  100.000000  100.000000  100.000000     100.000000   
2019-06-02   99.902422  102.129743  100.939179  102.089457     102.089457   
2019-06-03  101.958367  101.366855   96.729470   95.854498      95.854498   
2019-06-04   95.767888   95.193202   89.187279   90.001822      90.001822   
2019-06-05   89.858726   91.609263   90.415585   91.361703      91.361703   
...                ...         ...         ...         ...            ...   
2021-10-08  627.515157  648.337236  632.996052  630.169816     630.169816   
2021-10-09  629.003833  642.250296  633.551254  641.850959     641.850959   
2021-10-10  640.935932  653.882638  639.789636  639.554787     639.554787   
2021-10-11  638.385204  670.017566  642.802139  671.236310     671.236310   
2021-10-12  669.639980  665.621585  666.464175  665.698879     665.698879   

            Volume_BTC   Open_GLDM   High_GLDM    Low_GLDM  Close_GLDM  \
Date                                                                     
2019-06-01  100.000000         NaN         NaN         NaN         NaN   
2019-06-02   90.118919         NaN         NaN         NaN         NaN   
2019-06-03   97.848695  100.000000  100.000000  100.000000  100.000000   
2019-06-04  109.433473  100.609292  100.000000  100.533944  100.075586   
2019-06-05   96.765518  101.675552  100.829563  101.144165  100.302343   
...                ...         ...         ...         ...         ...   
2021-10-08  154.750997  134.729634  133.408756  133.150259  132.048367   
2021-10-09  144.480491  134.729634  133.408756  133.150259  132.048367   
2021-10-10  175.770450  134.729634  133.408756  133.150259  132.048367   
2021-10-11  189.597813  132.901759  131.960777  132.951945  131.746032   
2021-10-12  182.771206  132.901759  131.960777  132.951945  131.746032   

            Adj Close_GLDM  Volume_GLDM  
Date                                     
2019-06-01             NaN          NaN  
2019-06-02             NaN          NaN  
2019-06-03      100.000000   100.000000  
2019-06-04      100.075586    61.395157  
2019-06-05      100.302343    72.472967  
...                    ...          ...  
2021-10-08      132.048367   131.223554  
2021-10-09      132.048367   131.223554  
2021-10-10      132.048367   131.223554  
2021-10-11      131.746032    56.364598  
2021-10-12      131.746032    56.364598  

[865 rows x 12 columns]

In [62]:
# Backward fill the first 2 dates for GLDM
df[gldm_cols] = df[gldm_cols].fillna(method='bfill')

In [63]:
# Double checking the dataframe 
df

Open_BTC    High_BTC     Low_BTC   Close_BTC  Adj Close_BTC  \
Date                                                                        
2019-06-01  100.000000  100.000000  100.000000  100.000000     100.000000   
2019-06-02   99.902422  102.129743  100.939179  102.089457     102.089457   
2019-06-03  101.958367  101.366855   96.729470   95.854498      95.854498   
2019-06-04   95.767888   95.193202   89.187279   90.001822      90.001822   
2019-06-05   89.858726   91.609263   90.415585   91.361703      91.361703   
...                ...         ...         ...         ...            ...   
2021-10-08  627.515157  648.337236  632.996052  630.169816     630.169816   
2021-10-09  629.003833  642.250296  633.551254  641.850959     641.850959   
2021-10-10  640.935932  653.882638  639.789636  639.554787     639.554787   
2021-10-11  638.385204  670.017566  642.802139  671.236310     671.236310   
2021-10-12  669.639980  665.621585  666.464175  665.698879     665.698879   

            Volume_BTC   Open_GLDM   High_GLDM    Low_GLDM  Close_GLDM  \
Date                                                                     
2019-06-01  100.000000  100.000000  100.000000  100.000000  100.000000   
2019-06-02   90.118919  100.000000  100.000000  100.000000  100.000000   
2019-06-03   97.848695  100.000000  100.000000  100.000000  100.000000   
2019-06-04  109.433473  100.609292  100.000000  100.533944  100.075586   
2019-06-05   96.765518  101.675552  100.829563  101.144165  100.302343   
...                ...         ...         ...         ...         ...   
2021-10-08  154.750997  134.729634  133.408756  133.150259  132.048367   
2021-10-09  144.480491  134.729634  133.408756  133.150259  132.048367   
2021-10-10  175.770450  134.729634  133.408756  133.150259  132.048367   
2021-10-11  189.597813  132.901759  131.960777  132.951945  131.746032   
2021-10-12  182.771206  132.901759  131.960777  132.951945  131.746032   

            Adj Close_GLDM  Volume_GLDM  
Date                                     
2019-06-01      100.000000   100.000000  
2019-06-02      100.000000   100.000000  
2019-06-03      100.000000   100.000000  
2019-06-04      100.075586    61.395157  
2019-06-05      100.302343    72.472967  
...                    ...          ...  
2021-10-08      132.048367   131.223554  
2021-10-09      132.048367   131.223554  
2021-10-10      132.048367   131.223554  
2021-10-11      131.746032    56.364598  
2021-10-12      131.746032    56.364598  

[865 rows x 12 columns]

In [64]:
# Double checking NAs after both fillna functions
df.isnull().sum()

Open_BTC          4
High_BTC          4
Low_BTC           4
Close_BTC         4
Adj Close_BTC     4
Volume_BTC        4
Open_GLDM         0
High_GLDM         0
Low_GLDM          0
Close_GLDM        0
Adj Close_GLDM    0
Volume_GLDM       0
dtype: int64

In [65]:
# Checking overall df again
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 865 entries, 2019-06-01 to 2021-10-12
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Open_BTC        861 non-null    float64
 1   High_BTC        861 non-null    float64
 2   Low_BTC         861 non-null    float64
 3   Close_BTC       861 non-null    float64
 4   Adj Close_BTC   861 non-null    float64
 5   Volume_BTC      861 non-null    float64
 6   Open_GLDM       865 non-null    float64
 7   High_GLDM       865 non-null    float64
 8   Low_GLDM        865 non-null    float64
 9   Close_GLDM      865 non-null    float64
 10  Adj Close_GLDM  865 non-null    float64
 11  Volume_GLDM     865 non-null    float64
dtypes: float64(12)
memory usage: 120.1 KB


## Get dataframe for linechart 

In [66]:
adj_close_cols = [col for col in df.columns.tolist() if col.startswith('Adj Close')]
adj_close_cols

['Adj Close_BTC', 'Adj Close_GLDM']

In [67]:
linedata = df.loc[:, adj_close_cols]

In [68]:
# Checking the subsetted dataframe
linedata

Adj Close_BTC  Adj Close_GLDM
Date                                     
2019-06-01     100.000000      100.000000
2019-06-02     102.089457      100.000000
2019-06-03      95.854498      100.000000
2019-06-04      90.001822      100.075586
2019-06-05      91.361703      100.302343
...                   ...             ...
2021-10-08     630.169816      132.048367
2021-10-09     641.850959      132.048367
2021-10-10     639.554787      132.048367
2021-10-11     671.236310      131.746032
2021-10-12     665.698879      131.746032

[865 rows x 2 columns]

In [69]:
# Rename column names
linedata.columns = ['Bitcoin', 'GLDM']

In [70]:
linedata.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 865 entries, 2019-06-01 to 2021-10-12
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Bitcoin  861 non-null    float64
 1   GLDM     865 non-null    float64
dtypes: float64(2)
memory usage: 52.6 KB


In [71]:
# round floats
for col in linedata.columns.tolist():
    linedata[col] = round(linedata[col],0)

In [72]:
linedata.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 865 entries, 2019-06-01 to 2021-10-12
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Bitcoin  861 non-null    float64
 1   GLDM     865 non-null    float64
dtypes: float64(2)
memory usage: 52.6 KB


In [73]:
# Double checking the dataframe
linedata

Bitcoin   GLDM
Date                      
2019-06-01    100.0  100.0
2019-06-02    102.0  100.0
2019-06-03     96.0  100.0
2019-06-04     90.0  100.0
2019-06-05     91.0  100.0
...             ...    ...
2021-10-08    630.0  132.0
2021-10-09    642.0  132.0
2021-10-10    640.0  132.0
2021-10-11    671.0  132.0
2021-10-12    666.0  132.0

[865 rows x 2 columns]

## Load dataframe as excel file 

### Define ExcelChart class

In [74]:
class ExcelChart():
    def __init__(
        self,
        chart_df: pd.DataFrame,
        raw_df_list: list,
        start_date: str, 
        end_date: str,
        filename: str, 
        filepath: str
    ):
        """Create a new excel sheet with a line chart as the first sheet, 
        append raw dataframes as separate worksheets, 
        then save the excel workbook.

        Args:
            chart_df (pd.DataFrame): pandas DataFrame used to plot line chart of indexed price history data
            raw_df_list (list): list of raw and indexed dataframes of price data for each financial asset
            start_date (str): start date of line chart
            end_date (str): end date of line chart
            filename (str): file name of excel file
            filepath (str): absolute file path to save excel file to
        """
        self.chart_df = chart_df
        self.raw_df_list = raw_df_list
        self.start_date = start_date
        self.end_date = end_date
        self.filename = filename
        self.filepath = filepath
    
    ###########
    # Helpers
    ###########

    def format_date_display(
        self,
        start_date: str,
        end_date: str):
        """Converts start date and end date from DD/MM/YYYY format to dd month_name yy format for excel chart display.

        Args:
            start_date (str): start date in string in DD/MM/YYYY format
            end_date (str):  end date in string in DD/MM/YYYY format

        Returns:
            start_date (str): start date in string in dd month_name yy format
            end_date (str):  end date in string in dd month_name yy format
        """

        start_date = datetime.strptime(start_date, '%d/%m/%Y') # read str as datetime
        start_date = datetime.strftime(start_date, '%d %b %y') # convert to dd month_name yy format

        end_date = datetime.strptime(end_date, '%d/%m/%Y') # read str as datetime
        end_date = datetime.strftime(end_date, '%d %b %y') # convert to dd month_name yy format

        return start_date, end_date

    def new_excelsheet(
        self,
        df: pd.DataFrame, 
        sheetname: str, 
        sheetnumber: int = 0
    ):
        """Writes a pandas dataframe as a new excel spreadsheet in an excel workbook

        Args:
            df (pd.DataFrame): pandas dataframe to be written as an excel spreadsheet
            sheetname (str): name of new excel spreadsheet
            sheetnumber (int, optional): sheet number of the new excel spreadsheet. 
            Defaults to 0 (default sheet/first displayed sheet)

        Returns:
            wb (object): excel Workbook object
        """
        # remove timestamp from datetime index
        df.index = df.index.date 

        wb = Workbook()
        newsheet = wb.create_sheet(sheetname, sheetnumber)
        
        for row in dataframe_to_rows(df, index = True, header = True):
            newsheet.append(row)
            
        for cell in newsheet[1]: # excel indexing starts from 1, to select row 1
            cell.style = 'Pandas'
        
        # Drop default blank 'Sheet' in workbook
        wb.remove(wb['Sheet'])
        
        return wb

    def append_excelsheet(
        self,
        wb: object, 
        df: pd.DataFrame, 
        sheetname: str, 
        sheetnumber: int =0
    ):
        """Appends a pandas dataframe as an additional excel spreadsheet to an existing excel Workbook

        Args:
            wb (object): excel Workbook object 
            df (pd.DataFrame): pandas dataframe to append as a new excel spreadsheet
            sheetname (str): name of newly appended worksheet
            sheetnumber (int, optional): Sheet number of newly appended worksheet in excel workbook. 
            Defaults to 0 (first displayed worksheet).

        Returns:
            wb (object): excel workbook object with appended worksheet
        """
        # remove timestamp from datetime index
        df.index = df.index.date 
        
        newsheet = wb.create_sheet(sheetname, sheetnumber)
        
        for row in dataframe_to_rows(df, index = True, header = True):
            newsheet.append(row)
            
        for cell in newsheet[1]: # excel indexing starts from 1, to select row 1
            cell.style = 'Pandas'
        
        return wb

    def create_linechart(
        self,
        wb: object,
        title: str,
        ytitle: str,
        date_unit: str,
        date_format: str,
        xtitle: str = "Date",
        chartcell: str ='G3'
    ):
        """Creates a new time-series line chart from data wthin an excel workbook object 
        using Date as the first column.

        Args:
            wb (object): excel workbook object
            title (str): chart title
            ytitle (str): y axis title
            date_unit (str): major date units for x axis in 'days' or 'months' format.
            date_format (str): format for displayed date.
            xtitle (str, optional): x axis title. Defaults to "Date".
            chartcell (str, optional): cell in workbook object to place chart. Defaults to 'G3'.

        Returns:
            wb (object): excel workbook object with newly created linechart
        """
        
        sheet = wb.active
        
        # exclude date column when selecting data, include header
        data = Reference(
            worksheet=sheet,
            min_row=1,
            max_row=sheet.max_row,
            min_col=2,
            max_col=sheet.max_column
        )

        # select only date column without title
        dates = Reference(
            worksheet=sheet,
            min_col=1,
            max_col=1,
            min_row=2, 
            max_row=sheet.max_row
        )
        
        chart = LineChart()
        chart.title = title
        chart.style = 2
        
        chart.y_axis.title = ytitle
        chart.x_axis.title = xtitle
        
        chart.y_axis.crossAx = 500
        chart.x_axis = DateAxis(crossAx=100)
        chart.x_axis.number_format = date_format
        chart.x_axis.majorTimeUnit = date_unit
        
        chart.height = 12
        chart.width = 25
        
        chart.add_data(data, titles_from_data=True)
        
        chart.set_categories(dates)
        
        sheet.add_chart(chart, chartcell)

        return wb
    
    #################
    # Core function
    #################

    def run(self):
        """Create a new excel sheet with line chart, append raw dataframes as separate worksheets, 
        then save the excel workbook.
        """
        
        excel = self.new_excelsheet(
            df = self.chart_df, 
            sheetname = 'Chart', 
            sheetnumber = 0
        )

        start_date, end_date = self.format_date_display(
            self.start_date,
            self.end_date
        )

        excel = self.create_linechart(
            wb = excel,
            title = f"Indexed price history ({start_date} - {end_date})",
            ytitle = f"% Change from {start_date}",
            date_unit = "months",
            date_format = "mmm-yy"
        )

        for list_index, df in enumerate(self.raw_df_list):  

            excel = self.append_excelsheet(
                wb = excel, 
                df = df, 
                sheetname = df.name,
                sheetnumber = list_index + 1
            )

        excel.save(f'{self.filepath}/{self.filename}.xlsx')

        excel.close()

# Create new excel sheet with line chart, append raw dataframes as separate worksheets, then save the excel workbook

In [75]:
raw_dataframes = [bitcoin, bitcoin_indexed, gldm, gldm_indexed]

In [76]:
# set dataframe names
bitcoin.name = 'Bitcoin raw'
bitcoin_indexed.name = 'Bitcoin indexed'
gldm.name = 'GLDM raw'
gldm_indexed.name = 'GLDM indexed'

# get today's date 
TODAY = str(datetime.today().strftime('%d %b %y'))

excel_filepath = Path.cwd().parents[0]

In [77]:
create_excel_chart = ExcelChart(
        chart_df = linedata,
        raw_df_list = raw_dataframes,
        start_date = start_date, 
        end_date = end_date,
        filename = 'Historical_price-notebook_example', 
        filepath = excel_filepath
)
create_excel_chart.run()

# Further ideas

The above is just a simple example of how we can combine rpa with other python packages. 
Of course, web scraping can be more efficient using other python web scraping packages like [scrapy](https://scrapy.org/) or [beautiful soup](https://beautiful-soup-4.readthedocs.io/en/latest/). 

We can also convert this into a main pipeline script after tidying up some of the code. 
The following additional features can also be included:
- reading the required user-defined information from a configutation file. 
  - i.e. instead of only bitcoin and GLDM, we can include a host of other user-specified stocks to monitor  
- setting up an Simple Mail Transfer Protocol (SMTP) logging handler to trigger an email alert if a certain price threshold is reached  
- reading email passwords from Windows Credentials
- converting the main script into an executable file, and then using Windows Task Scheduler to trigger the download of files from the web and excel processing at specified times 